sycl/plugins/unified_runtime/ur/adapters/hip/enqueue.cpp

```diff
@@ -1328,20 +1384,99 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueUSMPrefetch(
+                      UR_RESULT_SUCCESS);
+      return UR_RESULT_ERROR_ADAPTER_SPECIFIC;
+    }
+
+    // TODO: If pMem points to valid system-allocated pageable memory, we should
+    // check that the device also has the hipDeviceAttributePageableMemoryAccess
+    // property, so that a valid read-only copy can be created on the device.
+    // This also applies for UR_USM_MEM_ADVICE_SET/MEM_ADVICE_CLEAR_READ_MOSTLY.
+  }
+
+  // NOTE: The hipPointerGetAttribute API is marked as beta, meaning, while this
+  // is feature complete, it is still open to changes and outstanding issues.
   unsigned int PointerRangeSize = 0;
-  UR_CHECK_ERROR(hipPointerGetAttribute(&PointerRangeSize,
-                                        HIP_POINTER_ATTRIBUTE_RANGE_SIZE,
-                                        (hipDeviceptr_t)HIPDevicePtr));
+  UR_CHECK_ERROR(hipPointerGetAttribute(
+      &PointerRangeSize, HIP_POINTER_ATTRIBUTE_RANGE_SIZE,
+      static_cast<hipDeviceptr_t>(HIPDevicePtr)));
   UR_ASSERT(size <= PointerRangeSize, UR_RESULT_ERROR_INVALID_SIZE);
 
-  // TODO implement a mapping to hipMemAdvise once the expected behaviour
-  // of urEnqueueUSMAdvise is detailed in the USM extension
-  return urEnqueueEventsWait(hQueue, 0, nullptr, phEvent);
+  ur_result_t Result = UR_RESULT_SUCCESS;
+  std::unique_ptr<ur_event_handle_t_> EventPtr{nullptr};
+
+  try {
+    ScopedContext Active(Device);
+
+    if (phEvent) {
+      EventPtr =
+          std::unique_ptr<ur_event_handle_t_>(ur_event_handle_t_::makeNative(
+              UR_COMMAND_USM_ADVISE, hQueue, hQueue->getNextTransferStream()));
+      EventPtr->start();
+    }
+
+    const auto DeviceID = Device->get();
+    if (advice & UR_USM_ADVICE_FLAG_DEFAULT) {
+      UR_CHECK_ERROR(
+          hipMemAdvise(pMem, size, hipMemAdviseUnsetReadMostly, DeviceID));
+      UR_CHECK_ERROR(hipMemAdvise(
+          pMem, size, hipMemAdviseUnsetPreferredLocation, DeviceID));
+      UR_CHECK_ERROR(
+          hipMemAdvise(pMem, size, hipMemAdviseUnsetAccessedBy, DeviceID));
+    } else {
+      Result = setHipMemAdvise(HIPDevicePtr, size, advice, DeviceID);
+      // UR_RESULT_ERROR_INVALID_ENUMERATION is returned when using a valid but
+      // currently unmapped advice arguments as not supported by this platform.
+      // Therefore, warn the user instead of throwing and aborting the runtime.
+      if (Result == UR_RESULT_ERROR_INVALID_ENUMERATION) {
+        setErrorMessage("mem_advise is ignored as the advice argument is not "
+                        " supported by this device.",
+                        Result);
+        return UR_RESULT_ERROR_ADAPTER_SPECIFIC;
+      }
+    }
+
+    if (phEvent) {
+      Result = EventPtr->record();
+      *phEvent = EventPtr.release();
+    }
+  } catch (ur_result_t err) {
+    Result = err;
+  } catch (...) {
+    Result = UR_RESULT_ERROR_UNKNOWN;
+  }
+
+  return Result;
 #else
   return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
```

Comment:
Since memory advices are effectively just hints, should we maybe just ignore them if we do not have capabilities to do something with them? Alternative is that users crash from this error despite us considering this implemented.

Reply:
@npmiller - Preferences here?

Reply:
Yeah I don't think we should error out if we don't support the advise, having a warning is good though.

Reply:
Okay thanks, settling on this then! Passing `UR_SUCCESS` errc with the message. Therefore, the handling of PI errors for `PI_ERROR_PLUGIN_SPECIFIC_ERROR` will just print a warning (if warning level >=2 is enabled) and `return;` without throwing.

Reply:
I think part of this discussion is also for the line above, i.e. we should set the "ignored" warning in the `#else` branch.

Reply:
@steffenlarsen Ah yeh sorry I didn't meantion this but I left it as is in this PR because @JackAKirk has already opened https://github.com/intel/llvm/pull/10958 that resolves this. Pretty much related to `hipPointerGetAttribute` requiring ROCM 5.0+, so Jack's PR only if-def's that API and removes the unsupported feature return error code. Let me know if you'd prefer me still adding any changes here. Thanks!

Reply:
Thank you for clarifying! That's perfect. 😄